lightning/src/ln/channelmanager.rs

```diff
@@ -1850,17 +1850,24 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 						break Some(("Forwarding node has tampered with the intended HTLC values or origin node has an obsolete cltv_expiry_delta", 0x1000 | 13, Some(self.get_channel_update_for_unicast(chan).unwrap())));
 					}
 					let cur_height = self.best_block.read().unwrap().height() + 1;
-					// Theoretically, channel counterparty shouldn't send us a HTLC expiring now, but we want to be robust wrt to counterparty
-					// packet sanitization (see HTLC_FAIL_BACK_BUFFER rational)
+					// Theoretically, channel counterparty shouldn't send us a HTLC expiring now,
+					// but we want to be robust wrt to counterparty packet sanitization (see
+					// HTLC_FAIL_BACK_BUFFER rationale).
 					if msg.cltv_expiry <= cur_height + HTLC_FAIL_BACK_BUFFER as u32 { // expiry_too_soon
 						break Some(("CLTV expiry is too close", 0x1000 | 14, Some(self.get_channel_update_for_unicast(chan).unwrap())));
 					}
 					if msg.cltv_expiry > cur_height + CLTV_FAR_FAR_AWAY as u32 { // expiry_too_far
 						break Some(("CLTV expiry is too far in the future", 21, None));
 					}
-					// In theory, we would be safe against unintentional channel-closure, if we only required a margin of LATENCY_GRACE_PERIOD_BLOCKS.
-					// But, to be safe against policy reception, we use a longer delay.
-					if (*outgoing_cltv_value) as u64 <= (cur_height + HTLC_FAIL_BACK_BUFFER) as u64 {
+					// If the HTLC expires ~now, don't bother trying to forward it to our
+					// counterparty. They should fail it anyway, but we don't want to bother with
+					// the round-trips or risk them deciding they definitely want the HTLC and
+					// force-closing to ensure they get it if we're offline.
+					// We previously had a much more aggressive check here which tried to ensure
+					// our counterparty receives an HTLC which has *our* risk threshold met on it,
+					// but there is no need to do that, and since we're a bit conservative with our
```

Comment:
I think it can be interesting to dissociate the two different types of risks we're exposed too. The first risk is missing the onchain confirmation of a HTLC on the backward link and as such accounting a loss in our forward HTLC balance. The second risk is the automatic force-close of a channel to achieve the onchain confirmation.

I think what this PR highlights is that we can decrease the latter without affecting the former. Onchain confirmation is function of when the forward link is going to go onchain (`should_broadcast_holder_commitment_tx`) and the CLTV delta between the HTLC on the backward link. None of those values are affected by this change.

However I think we increase our exposure to the second type of risk. In `should_broadcast_holder_commitment_tx`, we decide to go onchain only if an outbound HTLC is _behind_ current height by `LATENCY_GRACE_PERIOD_BLOCKS`. Previously I think we had a buffer of HTLC_FAIL_BACK_BUFFER + LATENCY_GRACE_PERIOD_BLOCKS in the worst-case scenario of forwarding a HTLC with an expiration at the boundary height value. Now, we only have `LATENCY_GRACE_PERIOD_BLOCKS` * 2.

I think this is okay, though if we observe an increase rate of channel force-closed, we could still bump to an intermediary value between `LATENCY_GRACE_PERIOD_BLOCKS` and `HTLC_FAIL_BACK_BUFFER`.

Reply:
Right, I think part of the argument in this PR is that we kinda leave correct handling up to our counterparty. If we give them an HTLC which expires soon they either need to claim it ASAP or fail it ASAP, and probably shouldn't themselves forward it. This is why we use `LATENCY_GRACE_PERIOD_BLOCKS` - its the "we may be desync'd from our peer by this many blocks, so give a bit of a grace period before we do anything drastic" constant.

If our counterparty *does* take a few blocks to fail/claim the HTLC and we have to force-close, I think that speaks more to our counterparty being buggy than us.

Reply:
Yes, the only cases of non-buggy peers taking more than few blocks to fail/claim the HTLC are hold-invoices or Lightning Rod styles of setup. Though their inflated CLTVs are going to be encompassed within `outgoing_cltv_value` and should not interfere with `LATENCY_GRACE_PERIOD_BLOCKS`.

Reply:
> Yes, the only cases of non-buggy peers taking more than few blocks to fail/claim the HTLC are hold-invoices or Lightning Rod styles of setup. Though their inflated CLTVs are going to be encompassed within outgoing_cltv_value and should not interfere with LATENCY_GRACE_PERIOD_BLOCKS.

I don't think that matters? By the time an invoice gets close to claimable, our counterparty needs to have failed the forwarding end on-chain so that they can do an offchain failure with us. If they don't or if they forward without enough time to do so, then they are buggy, long-CLTV or not.

Reply:
> I don't think that matters?

Yes we agree here, this is what I meant by "should not interfere" :)